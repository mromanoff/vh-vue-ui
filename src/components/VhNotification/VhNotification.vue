<template>
  <div
    class="VhNotification"
    :class="computedClasses"
  >
    <vh-icon
      class="VhNotification-svg"
      name="error"
      :color="computedIconColor"
      size="large"
    />
    <div class="VhNotification-message">
      <slot />
    </div>
    <button
      v-if="dismissible"
      class="VhNotification-closeButton"
    >
      <vh-icon
        name="close"
        size="medium"
        :color="computedIconColor"
      />
    </button>
  </div>
</template>

<script>
export default {
  name: 'VhNotification',
  props: {
    /**
     * Notification Type: info, warning, success, error.
     */
    type: {
      type: String,
      default: 'info',
      validator: value => {
        return value.match(/(info|warning|success|error)/);
      },
    },
    /**
     * Dismissible Notification: true, false
     */
    dismissible: {
      type: Boolean,
      default: false,
    },
    /**
     * VhNotification is active
     */
    value: {
      type: Boolean,
      default: true,
    },
  },

  computed: {
    computedClasses() {
      return {
        'VhNotification--info': this.type === 'info',
        'VhNotification--warning': this.type === 'warning',
        'VhNotification--success': this.type === 'success',
        'VhNotification--error': this.type === 'error',
      };
    },

    computedIconColor() {
      let color;
      switch (this.type) {
        case 'info':
          color = 'lightBlue';
          break;

        case 'warning':
          color = 'yellow';
          break;

        case 'success':
          color = 'green';
          break;

        case 'error':
          color = 'red';
          break;
        default:
          console.warn("Notification type doesn't exist " + this.type + '.');
      }

      return color;
    },
  },
  methods: {
    onClick() {
      /* TOTO: please verify this implementation */
      this.$emit('remove', this);
    },
  },
};
</script>

<style scoped>
.VhNotification {
  position: relative;
  display: flex;
  padding: var(--space--medium) 0;
  color: var(--text-color);
  align-items: center;
}

.VhNotification-svg {
  width: 72px;
  flex: 0 0 72px;
}

.VhNotification-message {
  flex: 1 1 auto;
  margin-right: 64px;
}

.VhNotification-closeButton {
  position: absolute;
  right: 8px;
  top: 8px;
  border: none;
  width: 44px;
  height: 44px;
  padding: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.VhNotification-closeButton:focus {
  outline: 0;
}

.VhNotification--info {
  background-color: color-mod(var(--color--blue200) a(50%));
  color: var(--color--blue700);
}

.VhNotification--success {
  background-color: color-mod(var(--color--green300) a(50%));
  color: var(--color--green700);
}

.VhNotification--warning {
  background-color: color-mod(var(--color--yellow300) a(50%));
  color: var(--color--yellow700);
}

.VhNotification--error {
  background-color: color-mod(var(--color--red300) a(50%));
  color: var(--color--red700);
}
</style>
