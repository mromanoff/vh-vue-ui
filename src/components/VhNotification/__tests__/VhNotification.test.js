import { shallowMount } from '@vue/test-utils';
import VhNotification from '../VhNotification.vue';

describe('VhNotification.vue', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(VhNotification, {
      propsData: {
        type: 'info',
      },
      slots: {
        default: 'A simple error notification message.',
      },
      stubs: ['vh-icon'],
    });
  });

  it('renders default Info VhNotification', () => {
    expect(wrapper.classes()).toStrictEqual(['VhNotification', 'VhNotification--info']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Warning VhNotification', () => {
    wrapper = shallowMount(VhNotification, {
      propsData: {
        type: 'warning',
      },
      slots: {
        default: 'A simple error notification message.',
      },
      stubs: ['vh-icon'],
    });

    expect(wrapper.classes()).toStrictEqual(['VhNotification', 'VhNotification--warning']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Success VhNotification', () => {
    wrapper = shallowMount(VhNotification, {
      propsData: {
        type: 'success',
      },
      slots: {
        default: 'A simple error notification message.',
      },
      stubs: ['vh-icon'],
    });

    expect(wrapper.classes()).toStrictEqual(['VhNotification', 'VhNotification--success']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Error VhNotification', () => {
    wrapper = shallowMount(VhNotification, {
      propsData: {
        type: 'error',
      },
      slots: {
        default: 'A simple error notification message.',
      },
      stubs: ['vh-icon'],
    });

    expect(wrapper.classes()).toStrictEqual(['VhNotification', 'VhNotification--error']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders dismissible Info VhNotification', () => {
    wrapper = shallowMount(VhNotification, {
      propsData: {
        type: 'info',
        dismissible: true,
      },
      slots: {
        default: 'A simple error notification message.',
      },
      stubs: ['vh-icon'],
    });

    const closeButton = wrapper.find('.VhNotification-closeButton');
    expect(wrapper.classes()).toStrictEqual(['VhNotification', 'VhNotification--info']);
    expect(closeButton.is('button')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
});
