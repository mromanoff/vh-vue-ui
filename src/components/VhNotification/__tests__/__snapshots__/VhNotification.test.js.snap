// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VhNotification.vue renders Error VhNotification 1`] = `
<div class="VhNotification VhNotification--error">
  <vh-icon-stub name="error" color="red" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div>
  <!---->
</div>
`;

exports[`VhNotification.vue renders Success VhNotification 1`] = `
<div class="VhNotification VhNotification--success">
  <vh-icon-stub name="error" color="green" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div>
  <!---->
</div>
`;

exports[`VhNotification.vue renders Warning VhNotification 1`] = `
<div class="VhNotification VhNotification--warning">
  <vh-icon-stub name="error" color="yellow" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div>
  <!---->
</div>
`;

exports[`VhNotification.vue renders default Info VhNotification 1`] = `
<div class="VhNotification VhNotification--info">
  <vh-icon-stub name="error" color="lightBlue" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div>
  <!---->
</div>
`;

exports[`VhNotification.vue renders dismissible Info VhNotification 1`] = `
<div class="VhNotification VhNotification--info">
  <vh-icon-stub name="error" color="lightBlue" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div> <button class="VhNotification-closeButton">
    <vh-icon-stub name="close" size="medium" color="lightBlue"></vh-icon-stub>
  </button>
</div>
`;

exports[`VhNotification.vue renders dissmissible Info VhNotification 1`] = `
<div class="VhNotification VhNotification--info">
  <vh-icon-stub name="error" color="lightBlue" size="large" class="VhNotification-svg"></vh-icon-stub>
  <div class="VhNotification-message">A simple error notification message.</div> <button class="VhNotification-closeButton">
    <vh-icon-stub name="close" size="medium" color="lightBlue"></vh-icon-stub>
  </button>
</div>
`;
